lib/rex/proto/ms_dtyp.rb

```diff
@@ -0,0 +1,226 @@
+        _, _, ia, sa = val.split('-', 4)
+        val = {}
+        val[:identifier_authority] = [ia.to_i].pack('Q>')[2..].bytes
+        val[:sub_authority] = sa.split('-').map(&:to_i)
+      end
+
+      super
+    end
+
+    def to_s
+      str = 'S-1'
+      str << "-#{("\x00\x00" + identifier_authority.to_binary_s).unpack1('Q>')}"
+      str << '-' + sub_authority.map(&:to_s).join('-')
+    end
+  end
+
+  # [2.4.4.1 ACE_HEADER](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/628ebb1d-c509-4ea0-a10f-77ef97ca4586)
+  class MsDtypAceHeader < BinData::Record
+    endian :little
+
+    uint8  :ace_type
+    struct :ace_flags do
+      bit1 :failed_access_ace_flag
+      bit1 :successful_access_ace_flag
+      bit1 :reserved
+      bit1 :inherited_ace
+      bit1 :inherit_only_ace
+      bit1 :no_propagate_inherit_ace
+      bit1 :container_inherit_ace
+      bit1 :object_inherit_ace
+    end
+    uint16 :ace_size, initial_value: -> { parent&.num_bytes || 0 }
+  end
+
+  # [2.4.4.2 ACCESS_ALLOWED_ACE](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/72e7c7ea-bc02-4c74-a619-818a16bf6adb)
+  class MsDtypAccessAllowedAceBody < BinData::Record
+    endian :little
+
+    ms_dtyp_access_mask :access_mask
+    ms_dtyp_sid         :sid
+  end
+
+  # [2.4.4.2 ACCESS_ALLOWED_ACE](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/72e7c7ea-bc02-4c74-a619-818a16bf6adb)
+  class MsDtypAccessAllowedAce < BinData::Record
+    endian :little
+
+    ms_dtyp_ace_header              :header, initial_value: { ace_type: 0 }
+    ms_dtyp_access_allowed_ace_body :body
+  end
+
+  class MsDtypAce < BinData::Record
+    endian :little
+
+    ms_dtyp_ace_header :header
+    choice             :body, selection: -> { header.ace_type } do
+      ms_dtyp_access_allowed_ace_body 0
+      string                          :default, read_length: -> { header.ace_size - body.rel_offset }
+    end
+  end
+
+  # [2.4.5 ACL](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/20233ed8-a6c6-4097-aafa-dd545ed24428)
+  class MsDtypAcl < BinData::Record
+    ACL_REVISION = 2
+    ACL_REVISION_DS = 4
+
+    endian :little
+
+    uint8  :acl_revision, initial_value: ACL_REVISION
+    uint8  :sbz1
+    uint16 :acl_size, initial_value: -> { num_bytes }
+    uint16 :acl_count, initial_value: -> { aces.length }
+    uint16 :sbz2
+    array  :aces, type: :ms_dtyp_ace, initial_length: :acl_count
+  end
+
+  # [2.4.6 SECURITY_DESCRIPTOR](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/7d4dac05-9cef-4563-a058-f108abecce1d)
+  class MsDtypSecurityDescriptor < BinData::Record
+    endian :little
+
+    uint8  :revision, initial_value: 1
+    uint8  :sbz1
+    struct :control do
+      bit1 :ss
+      bit1 :dt
+      bit1 :sd
+      bit1 :sp, initial_value: -> { sacl? ? 1 : 0 }
+      bit1 :dd
+      bit1 :dp, initial_value: -> { dacl? ? 1 : 0 }
+      bit1 :gd
+      bit1 :od
```

Comment:
Not a blocker; Any thoughts on renaming these to be the expanded versions from the linked table in the spec?

```suggestion
      bit1 :ss
      bit1 :dt
      bit1 :sd
      bit1 :sp, initial_value: -> { sacl? ? 1 : 0 }
      bit1 :dd
      bit1 :dp, initial_value: -> { dacl? ? 1 : 0 }
      bit1 :gd
      bit1 :od
```

Reply:
As I read the spec, that would replace the field names with those of the constants. As it is now, the fields are named as the are defined in the spec.